atom/browser/net/atom_network_delegate.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#ifndef ATOM_BROWSER_NET_ATOM_NETWORK_DELEGATE_H_
+#define ATOM_BROWSER_NET_ATOM_NETWORK_DELEGATE_H_
+
+#include <map>
+#include <set>
+#include <string>
+
+#include "brightray/browser/network_delegate.h"
+#include "base/callback.h"
+#include "base/values.h"
+#include "extensions/common/url_pattern.h"
+#include "net/base/net_errors.h"
+#include "net/http/http_request_headers.h"
+#include "net/http/http_response_headers.h"
+
+namespace extensions {
+class URLPattern;
+}
+
+namespace atom {
+
+class AtomNetworkDelegate : public brightray::NetworkDelegate {
+ public:
+  struct BlockingResponse;
+  using Listener =
+      base::Callback<BlockingResponse(const base::DictionaryValue&)>;
+
+  enum EventTypes {
+    kInvalidEvent = 0,
+    kOnBeforeRequest = 1 << 0,
+    kOnBeforeSendHeaders = 1 << 1,
+    kOnSendHeaders = 1 << 2,
+    kOnHeadersReceived = 1 << 3,
+    kOnBeforeRedirect = 1 << 4,
+    kOnResponseStarted = 1 << 5,
+    kOnCompleted = 1 << 6,
+    kOnErrorOccurred = 1 << 7,
+  };
+
+  struct ListenerInfo {
+    ListenerInfo() {}
+    ~ListenerInfo() {}
+
+    std::set<extensions::URLPattern> url_patterns;
+    AtomNetworkDelegate::Listener callback;
+  };
+
+  struct BlockingResponse {
+    BlockingResponse() {}
+    ~BlockingResponse() {}
+
+    int Cancel() const {
+      return cancel ? net::ERR_BLOCKED_BY_CLIENT : net::OK;
+    }
+
+    bool cancel;
+    GURL redirectURL;
```

Comment:
C++ code should use `redirect_url `, the same rule also applies for `requestHeaders` and `responseHeaders`.